Valigator.Core/Data.cs

```diff
@@ -117,5 +120,35 @@ private static IDataContainer<TValue> GetErrorDataContainer(IDataContainer<TValu
 
 		public static implicit operator TValue(Data<TValue> data)
 			=> data.Value;
+
+		public override bool Equals(object obj) 
+			=> obj is Data<TValue> data 
+				&& this == data;
+
+		public static bool operator ==(Data<TValue> x, Data<TValue> y) 
+			=> x.DataDescriptor.Equals(y.DataDescriptor)
+				&& (x._value.Equals(y._value) 
+					|| (typeof(TValue).IsArray && StructuralComparisons.StructuralEqualityComparer.Equals(x._value, y._value)));	
+
+		public static bool operator !=(Data<TValue> x, Data<TValue> y)
+			=> !x._value.Equals(y._value);
+
+		public override int GetHashCode()
+		{
+			int hashCode = 943777100;
+			var typeParameter = typeof(TValue);
+
+			if (typeParameter.IsArray || (typeParameter.IsGenericType && typeParameter.GetInterfaces().Any(t => t.GetGenericTypeDefinition() == typeof(IEnumerable<>))))
+			{
+				hashCode = hashCode * -1521134295 + (_value as IStructuralEquatable).GetHashCode((IEqualityComparer)typeof(EqualityComparer<>).MakeGenericType(typeParameter.GetElementType()).GetProperty(nameof(EqualityComparer<int>.Default)).GetValue(this));
```

Comment:
@JohannesMoersch should we cache the creation of the generic `EqualityComparer`? My gut instinct is yes, but I know you have some design decisions around keeping memory footprints small that I'm not sure if this would violate.

Reply:
Scratch that, it looks like `IEnumerable` does implement `IStructuralComparable` so this whole generic type thing won't work. It'll have to compare each element manually.